api/src/neo4j/queries/search/modelSearch.js

```diff
@@ -167,48 +211,49 @@ const modelSearch = async ({ searchTerm, model, version, limit }) => {
 
 /*
  * The search consists of two steps
- * 1. Do a fuzzy search over all nodes covered by full-text search index
- * 2. Fetch results for each component type (parallelly) and return result
+ * 1. Do a fuzzy search over all nodes for each component type (parallelly) and
+ *    fetch unique IDs
+ * 2. Fetch detailed results for each component type (parallelly) given the
+ *    unique IDs obtained in the first step
  */
 const search = async ({ searchTerm, model, version, limit, includeCounts }) => {
   const v = version ? `:V${version}` : '';
 
   const term = sanitizeSearchString(searchTerm, true);
-
-  // The search term is used twice, once with exact match and once with
-  // fuzzy match. This seems to produce optimal results.
-  let statement = `
-CALL db.index.fulltext.queryNodes("fulltext", "${term} ${term}~")
-YIELD node, score
-WITH node, score, LABELS(node) as labelList
-OPTIONAL MATCH (node)-[${v}]-(parentNode:${model})
-WHERE node:${model} OR parentNode:${model}
-WITH DISTINCT(
-	CASE
-                WHEN EXISTS(node.id) AND NOT apoc.coll.intersection(${JSON.stringify(
-                  CHILD_LABELS
-                )}, LABELS(node))
-                THEN { id: node.id, labels: labelList, score: score }
-		ELSE { id: parentNode.id, labels: LABELS(parentNode), score: score }
-	END
-) as r 
-WHERE any(r IN r.labels WHERE r="${model}")
-RETURN r
-`;
-  if (limit) {
-    statement += `
-LIMIT ${limit}
-`;
+  if (term.length === 0) {
+    throw new Error(`Empty searchTerm!`);
   }
 
-  const results = await queryListResult(statement);
-
-  const uniqueIds = {};
-  for (let [pos, node] of Object.entries(results)) {
-    if (!(node['id'] in uniqueIds)) {
-      uniqueIds[node['id']] = node;
-    }
+  const resultFetchIds = [];
+  for (const component of [
+    'Metabolite',
+    'CompartmentalizedMetabolite',
+    'Gene',
+    'Reaction',
+    'Subsystem',
+    'Compartment',
+  ]) {
+    resultFetchIds.push(
+      fetchIds({ component, term, model, version: v, limit })
+    );
```

Comment:
A few points here.

1. The name `resultFetchIds` could perhaps be renamed to something like `idSearchQueries`. `result` could often be used to represent the result of a `Promise` as opposed to the query for the `Promise` itself.
2. The list from lines 228 to 235 seems to have the same contents of `COMPONENT_TYPES`, which we have already defined in another file.
3. This is a personal preference but I think a `map` would be more concise and look better than the `for` loop here.

Together these points could be refactored into the following.

```javascript
  const idSearchQueries = COMPONENT_TYPES.map(component =>
    fetchIds({ component, term, model, version: v, limit })
  );
```